[2단계 - 웹 자동차 경주] 오션(김동해) 미션 제출합니다.

donghae-kim opened the pull request:
안녕하세요 파즈! 오션🌊입니다! 2단계도 잘 부탁드립니다!

이번 2단계를 진행하면서 든 궁금증 입니다!
---
### Dao 에서 반환 타입
이번에 Dao에서 값을 반환함에 있어서 DB테이블과 일치하는 Entity를 만들어서 사용하였습니다. 이를 서비스 단으로 전달하고 서비스단에서 DTO를 변환하는 방식으로 진행하였습니다.
여기서 만약 Join을 한다면 Dao에서 Join한 값에 매핑되는 무언가(Entity같은)가 있어야 할 것 같은데 이것이 결국 Entity와 같은 역할을 하는 DTO일 것이라고 생각을 하였습니다. 그래서 제가 내린 결론은 Entity는 Dto인데, 영속성 계층에서 서비스 계층으로 전송되는 DTO를 Entity라고 생각하게 되었습니다. 이 생각이 옳은지 궁금합니다!

---
### JOIN시 쿼리 관리
이번에 JOIN을 하지는 않았지만 join을 할 경우에 어떤 Dao에서 처리되어야 하는지에 대해 잘 모르겠습니다. 

---
### DB에서 Winner관리 테이블
DB에 Winner Column을 삭제했습니다. (반란군이 되보고 싶었습니다)
이유는 Play_Result에 Winner를 둔다면 Car_Result에 값과 중복되는 값이 저장된다고 생각하였습니다. 
그래서 Car_Result에 새로운 is_win column을 생각해보았습니다. 하지만 소수의 win 체크를 위해 쓸모없는 데이터가 많이 삽입되어야 한다고 생각했습니다.(ex. 10명중 승자가 1명일 경우 1명을 위해 9명의 쓸모없는 데이터 삽입)
그래서 결국 Winner Column을 삭제하였고 이는 자바 코드로 Position을 통해 계산하였습니다.

감사합니다. 2단계 리뷰도 잘 부탁들비니다 !!

donghae-kim committed:
docs: 2단계 요구사항 추가

donghae-kim committed:
feat: Entity 추가

CarEntity, GameEntity 추가

donghae-kim committed:
feat: ID로 CAR_RESULT 값 조회 추가

findAllById

donghae-kim committed:
feat: PLAY_RESULT 값 조회 추가

findAll

donghae-kim committed:
feat: 이력조회 결과 반환 DTO 추가

GameRecordResponseDto

donghae-kim committed:
feat: 이력조회 결과 서비스 추가

donghae-kim committed:
feat: getGameRecord Get 메소드 추가

donghae-kim committed:
style: Console Application 및 Console Controller 이름 변경

donghae-kim committed:
refactor: console application 중간과정 로직 삭제 및 우승자와 player 별 최종 이동거리를 출력 하도록 수정

donghae-kim committed:
feat: 하나의 비즈니스 로직을 갖는 객체 생성

donghae-kim committed:
refactor: 콘솔,웹 중복 코드 제거

donghae-kim committed:
test: test코드 추가

donghae-kim committed:
refactor: Winner Column 삭제

donghae-kim committed:
style: 람다식 변수 수정

donghae-kim committed:
docs: docs 업데이트

donghae-kim committed:
fix: test 코드 에러 수정

Be-poz commented on src/main/java/racingcar/dto/GameRecordResponseDto.java:
```diff
@@ -0,0 +1,33 @@
+        this.racingCars = racingCars;
+    }
+
+    private List<String> convert(List<CarEntity> winners) {
```
무엇에서 무엇으로 convert 되는지 명시된 이름이면 더 좋을 것 같아요!

donghae-kim replied:
DTO에서 convert 되는 메소드 명을 조금 더 가독성있게 수정해보았습니다!

Be-poz commented on src/main/java/racingcar/controller/WebController.java:
```diff
@@ -26,4 +29,10 @@ public ResponseEntity<RacingGameResponseDto> playGame(final @RequestBody @Valid
         return ResponseEntity.ok(racingGameResponseDto);
     }
+
+    @GetMapping("/plays")
+    public ResponseEntity<List<GameRecordResponseDto>> getGameRecord(){
```
ResponseEntity가 ``@ResponseBody``의 역할을 해주긴 하지만 일반적으로 api에 대한 컨트롤러인 경우 어노테이션을 ``@Controller``대신 ``@RestController`` 를 사용해서 뷰 리졸빙을 위한 컨트롤러와 구분을 지어줍니다.  

현재 코드를 바꿀 필요는 없습니다! 일반적인 경우를 집고 넘어가고 싶어서 언급했습니다.

donghae-kim replied:
단순하게 ResponseBody의 차이라고 생각했고 api 컨트롤러와 뷰 리졸빙을 위한 컨트롤러 구별에 대한 생각을 못했던 것 같습니다!
명확하게 하기 위해 수정하였습니다~!

Be-poz commented on src/main/java/racingcar/dao/entity/CarEntity.java:
```diff
@@ -0,0 +1,25 @@
+package racingcar.dao.entity;
+
+public class CarEntity {
+    private final int play_result_id;
```
카멜케이스로 변경을 하면 좋을 것 같습니다. 카멜케이스로 변경을 해도 문제되지 않아보이네요.

donghae-kim replied:
DB와 매핑된다고 생각하여 그대로 사용했었습니다!
수정하였습니다

Be-poz commented on src/main/java/racingcar/dao/entity/CarEntity.java:
```diff
@@ -0,0 +1,25 @@
+
+public class CarEntity {
+    private final int play_result_id;
+    private final String name;
+    private final int position;
```
지금 다시보니깐 car_result 테이블의 column도 ``car_`` prefix를 제외한 name, position으로 가도될 것 같습니다.  
테이블 명으로 어떤 도메인에 대한 name, position인지 알 수 있으니까요.

Be-poz commented on src/main/java/racingcar/dao/InMemoryCarDao.java:
```diff
@@ -19,20 +20,32 @@ public InMemoryCarDao(JdbcTemplate jdbcTemplate) {
 
     @Override
-    public void saveAll(int gameId, List<Car> cars) {
+    public void saveAll(int gameId, List<Car> cars, List<Car> winners) {
```
반란군으로 가던 길의 흔적 인가요 .. ?! ``carDao.saveAll(gameId, cars.getCars(), winners);``

donghae-kim replied:
반란군이 반란을 잘 못했네요 ..😭
관련 코드 수정을 제대로 하지 못한 것 같아 수정하였습니다.

Be-poz commented on src/main/java/racingcar/service/GameService.java:
```diff
@@ -28,14 +37,26 @@ public RacingGameResponseDto executeRacingGame(RacingGameRequestDto racingGameRe
         return new RacingGameResponseDto(cars.getWinners(), cars.getCars());
     }
+
+    public List<GameRecordResponseDto> getGameRecord() {
+        List<GameEntity> gameEntities = gameDao.findAll();
+        return gameEntities.stream().map(gameEntity -> {
+            List<CarEntity> carEntities = carDao.findAllById(gameEntity.getId());
+            int max = carEntities.stream()
+                    .max(Comparator.comparing(CarEntity::getPosition)).orElseThrow().getPosition();
+
+            List<CarEntity> result = carEntities.stream().filter(car -> car.getPosition() == max).collect(Collectors.toList());
+
+            return new GameRecordResponseDto(result, carEntities);
+        }).collect(Collectors.toUnmodifiableList());
+    }
```
가독성이 너무 안좋습니다. 메서드로 따로 빼서 정리할 수 있을 것 같고, ``.``이 한 줄에 2개이상 오지 않도록 적절하게 체이닝 하게끔 변경해주세요! 웹 미션이 되었어도 레벨 1의 기본을 잊으시면 안됩니다!!!

donghae-kim replied:
기능별로 분리하여 가독성이 좋게 하였습니다!
레벨 1의 기본을 잊지 않겠습니다 ....😂

Be-poz commented on src/main/java/racingcar/dto/RacingGameResponseDto.java:
```diff
@@ -10,11 +10,19 @@ public class RacingGameResponseDto {
         this.racingCars = convert(racingCars);
     }
 
+    private List<String> convertWinner(List<Car> winners){
```
이 메서드의 이름 또한 마찬가지입니다.

Be-poz commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -2,33 +2,31 @@
-
-    public void printResult() {
-        System.out.println(PRINT_RESULT);
+    public void printWinner(List<Car> winner) {
+        List<String> winners = new ArrayList<>();
+        for(Car car : winner){
+            winners.add(car.getName());
+        }
+        System.out.printf(PRINT_WINNER, NameFormatConverter.joinNameWithDelimiter(winners));
```
printResult와 printWinner 메서드를 1개의 메서드로 만들고 ``RacingGameResponseDto`` 를 이용해서 출력하게끔 변경해봅시다!

donghae-kim replied:
console과 web 모두에서 같은 `ResponseDto`를 이용하여 뷰를 출력할 수 있도록 하였습니다!

Be-poz commented:
> 이번에 Dao에서 값을 반환함에 있어서 DB테이블과 일치하는 Entity를 만들어서 사용하였습니다. 이를 서비스 단으로 전달하고 서비스단에서 DTO를 변환하는 방식으로 진행하였습니다.
> 여기서 만약 Join을 한다면 Dao에서 Join한 값에 매핑되는 무언가(Entity같은)가 있어야 할 것 같은데 이것이 결국 Entity와 같은 역할을 하는 DTO일 것이라고 생각을 하였습니다. 그래서 제가 내린 결론은 Entity는 Dto인데, 영속성 계층에서 서비스 계층으로 전송되는 DTO를 Entity라고 생각하게 되었습니다. 이 생각이 옳은지 궁금합니다!

네 그렇게 생각하시면 이해하기 좋을 것 같습니다~! 저는 이것을 DTO라고 보기도 하고 VO로 보기도 합니다. 사실 단어가 뭐가 중요하겠나 라는 생각이 제일 크게 들긴합니다 ㅎㅎ  



> 이번에 JOIN을 하지는 않았지만 join을 할 경우에 어떤 Dao에서 처리되어야 하는지에 대해 잘 모르겠습니다.

해당 join을 하여 얻는 정보가 어느쪽 도메인의 정보에 초점이 맞춰져있는지를 보고 결정할 것 같습니다. 하지만, 이것 또한 케이스에 따라 다를 것 같습니다. 현재 미션에서는 테이블의 컬럼이 적고 하지만 정말 큰 테이블 끼리 조인을 하게되면(사실 테이블이 클 수록 조인을 하는 경우 보다는 따로 단일 테이블에 쿼리를 날리는 경우가 더 많습니다) 쿼리 성능을 위해 규모가 작은 테이블을 메인 테이블로 둘 것 같습니다. 쿼리 튜닝에 대한 내용은 레벨4에서 아마 배우실 수 있으실 겁니다.

Be-poz reviewed (changes requested):
안녕하세요 오션 2단계 코드 잘 봤습니다!  
간단한 미션이라 크게 코멘트 남긴 것은 없고 자잘하게 남겼으니 확인해주시면 감사하겠습니다.  
질문에 대한 답변 또한 별도의 코멘트로 남겼으니 확인해주세요.

donghae-kim committed:
style: 메소드 명 수정

convert -> convertWinnersName,
convertWinner -> convertWinnersName,
convert -> convertRacingCars

donghae-kim committed:
refactor: WebController @RestController 수정

donghae-kim committed:
refactor: Winner 관련 제거못한 코드 수정

donghae-kim committed:
style: CarEntity field 이름 변경

donghae-kim committed:
refactor: data.sql column 수정에 따른 Dao 수정

donghae-kim committed:
refactor: gameService 리팩터링

donghae-kim committed:
refactor: 콘솔에서 RacingGameResponseDto를 사용하여 출력하도록 수정

donghae-kim committed:
style: Location -> Position 이름 변경

donghae-kim committed:
test: test 코드 추가

donghae-kim commented:
안녕하세요 파즈! 오션🌊입니다! 

파즈가 언급해주신 부분과 부족하다고 느꼈던 부분을 리팩토링 하였습니다.
또한 테스트에 대한 요구사항은 없었지만 mock을 한번 사용해보고 싶어서 테스트 코드를 작성해보았습니다!

감사합니다 파즈!

donghae-kim committed:
refactor: 콘솔 리팩터링

Be-poz commented on src/test/java/racingcar/service/GameServiceTest.java:
```diff
@@ -0,0 +1,57 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class GameServiceTest {
+
+    @Mock
+    private NumberGenerator numberGenerator;
+    @Mock
+    private GameDao gameDao;
+    @Mock
+    private CarDao carDao;
+
+    @InjectMocks
+    private GameService gameService;
```
👍👍👍

Be-poz commented on src/main/java/racingcar/service/GameService.java:
```diff
@@ -28,14 +36,36 @@ public RacingGameResponseDto executeRacingGame(RacingGameRequestDto racingGameRe
         return new RacingGameResponseDto(cars.getWinners(), cars.getCars());
     }
+
+    public List<GameRecordResponseDto> getGameRecord() {
+        List<GameEntity> gameEntities = gameDao.findAll();
+
+        return gameEntities.stream().map(gameEntity -> {
+            List<CarEntity> carEntities = carDao.findAllById(gameEntity.getId());
+
+            return new GameRecordResponseDto(getWinners(carEntities), carEntities);
+        }).collect(Collectors.toUnmodifiableList());
+    }
+
+    private List<CarEntity> getWinners(List<CarEntity> carEntities) {
+        return carEntities.stream()
+                .filter(car -> car.getPosition() == getMaxPosition(carEntities))
+                .collect(Collectors.toList());
+    }
+
+    private int getMaxPosition(List<CarEntity> carEntities) {
+        CarEntity carEntity = carEntities.stream()
+                .max(Comparator.comparing(CarEntity::getPosition))
+                .orElseThrow();
+
+        return carEntity.getPosition();
+    }
```
😎

Be-poz reviewed (approved):
안녕하세요 오션 피드백 반영해주신 것 잘 확인했습니다. 테스트 추가하신 것도 아주 잘 하신 것 같습니다. 😆
이번 미션은 이만 머지하도록 하겠습니다. 레벨2 첫 미션 수고하셨습니다!
